### PR TITLE
fix: Set a unique savedMetricId when adding a new saved metric to controls

### LINKED ISSUE #12611: [explore]Non-metric uniqueness in Explore
It seems like metrics in explorer and keyed by name, however for charts which support multiple metrics, Superset cannot differentiate between selected saved metrics. This means if you add the same metric multiple times and change one of the metrics all references to said metric change. 

### Expected results

Either:

1. The same metric should not be able to be added multiple times (ensuring uniqueness).
2. Each selected metric should have a unique ID allow.

### Actual results

what actually happens.

#### How to reproduce the bug

1. Go to Explorer.
2. Choose a chart type which supports multiple metrics.
2. Add a saved metric.
3. Re-add the same saved metric.
4. Change one of the saved metrics.
5. Notice that both metrics changed.

### Environment

(please complete the following information):

- superset version: `superset version`
- python version: `python --version`
- node.js version: `node -v`

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
Saved metrics added to controls did not have a unique id - we distinguished them only by their names. If we added the same saved metric twice and then tried to edit 1, both items were being edited, because their names were equal.
This PR introduces a `savedMetricId`, which is added to a metric when it's added to controls via the metrics popover.
The saved metrics that had been added before this PR will not have this param until they are removed and added again, however everything should still work as expected.
**_The only case when the bug will still happen is if 2 or more exactly the same saved metrics are already in controls panel before this change is introduced. In such case, removing the metric and adding it again in control panel will fix the issue._**

Fixes https://github.com/apache/superset/issues/12611
Closes https://github.com/apache/superset/issues/12611
### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/15073128/105178344-2c995d80-5b28-11eb-88e3-2e55f0f4a63a.mov

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes https://github.com/apache/superset/issues/12611
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

CC: @junlincc @rusackas  @john-bodley @adam-stasiak

### CODE REVIEW

**Comment by kgabryje** [comment]
Converted to draft, because I noticed a bug - chart is not rendering when only a single metric is allowed. I'll investigate and reopen when it's ready

File: superset-frontend/src/explore/components/controls/MetricControl/AdhocMetricEditPopover.jsx [hunk 1]
```diff
@@ -125,7 +125,15 @@ export default class AdhocMetricEditPopover extends React.Component {
       label = metricLabel;
     }
 
-    const metric = savedMetric?.metric_name ? savedMetric : adhocMetric;
+    // if saved metric, generate a unique savedMetricId
+    const metric = savedMetric?.metric_name
+      ? {
+          ...savedMetric,
+          savedMetricId: `${Math.random()
```
Comment (john-bodley) [inline superset-frontend/src/explore/components/controls/MetricControl/AdhocMetricEditPopover.jsx:132]: This unique ID seems somewhat hard to grok. Why not use a UUID?

**Comment by ktmud** [comment]
We should probably just remove the added save metric from the options list instead (which I believe is also the original behavior before the control redesign?).

**Comment by adam-stasiak** [comment]
I saw in console warning related to this PR:
To reproduce:
Add metric
Set Custom name for this
Save chart and revisit page 

```Warning: Failed prop type: The prop `savedMetric.metric_name` is marked as required in `AdhocMetricOption`, but its value is `undefined`.
    in AdhocMetricOption (at MetricDefinitionValue.jsx:80)
    in MetricDefinitionValue (at MetricsControl.jsx:126)
    in div (created by Context.Consumer)
    in Styled(div) (at MetricsControl.jsx:380)
    in div (at MetricsControl.jsx:363)
    in MetricsControl (created by Context.Consumer)
    in WithTheme(MetricsControl) (created by DragDropContext(WithTheme(MetricsControl)))
    in DragDropContext(WithTheme(MetricsControl)) (at Control.tsx:42)
    in div (at Control.tsx:41)
    in Control (at ControlPanelsContainer.jsx:125)
    in div (at ControlRow.jsx:33)
    in div (at ControlRow.jsx:31)
    in ControlSetRow (at ControlPanelsContainer.jsx:182)
    in div (created by PanelBody)
    in PanelBody (at ControlPanelSection.jsx:109)
    in div (created by PanelCollapse)
    in Transition (created by Collapse)
    in Collapse (created by PanelCollapse)
    in PanelCollapse (at ControlPanelSection.jsx:108)
    in div (created by Panel)
    in Panel (created by Uncontrolled(Panel))
    in Uncontrolled(Panel) (created by ForwardRef)
    in ForwardRef (at ControlPanelSection.jsx:100)
    in ControlPanelSection (at ControlPanelsContainer.jsx:149)
    in div (created by TabPane)
    in TabPane (created by Context.Consumer)
    in Styled(TabPane) (at ControlPanelsContainer.jsx:245)
    in div (created by TabPanelList)
    in div (created by TabPanelList)
    in TabPanelList (created by ForwardRef(Tabs))
    in div (created by ForwardRef(Tabs))
    in ForwardRef(Tabs) (created by Tabs)
    in Tabs (created by Context.Consumer)
    in Styled(Tabs) (at ControlPanelsContainer.jsx:240)
    in div (created by Context.Consumer)
    in Styled(div) (at ControlPanelsContainer.jsx:226)
    in ControlPanelsContainer (created by ConnectFunction)
    in ConnectFunction (at ExploreViewContainer.jsx:475)
    in div (created by Resizable)
    in Resizable (at ExploreViewContainer.jsx:458)
    in div (created by Context.Consumer)
    in Styled(div) (at ExploreViewContainer.jsx:373)
    in ExploreViewContainer (created by ConnectFunction)
    in ConnectFunction (at App.jsx:38)
    in DynamicPluginProvider (at App.jsx:37)
    in ThemeProvider (at App.jsx:36)
    in Provider (at App.jsx:35)
    in App (created by HotExportedApp)
    in AppContainer (created by HotExportedApp)
    in HotExportedApp (at explore/index.jsx:59)```

**Comment by villebro** [comment]
@ktmud I agree, I don't see the value in adding an identical metric twice. In addition to causing unnecessary confusion for users it will cause problems in the backend where column names are assumed to be unique.

**Comment by junlincc** [comment]
I'm glad that you guys are all aligned, I originally thought we should not restrict users from adding the same metrics twice especially in table viz. well........ since there's no conflict between adding unique ID and removing same metric, should we move forward with this solution for now? @villebro

**Comment by ktmud** [comment]
We shall not need current solution in this PR. It adds unnecessary complexity to the code.

File: superset-frontend/src/explore/components/controls/MetricControl/AdhocMetricEditPopover.jsx (hunk 1 above)
Comment (kgabryje) [reply superset-frontend/src/explore/components/controls/MetricControl/AdhocMetricEditPopover.jsx:132]: I did not want to add another dependency to the project and that method of generating ID was already in use in `AdhocMetric.js`

**Comment by kgabryje** [comment]
I assumed that we should keep the possibility of adding the same metric twice. If that's not the case, simply removing added metrics from options in popover seems more sensible. I'll refactor it.

**Comment by kgabryje** [comment]
I'm closing this PR in favour of https://github.com/apache/superset/pull/12657, which implements removing added metrics from options in popover.

**Comment by junlincc** [comment]
> I assumed that we should keep the possibility of adding the same metric twice. If that's not the case, simply removing added metrics from options in popover seems more sensible. I'll refactor it.

That's my assumption too. i'm sure we will hear that business need someday. thanks for the PR though 🙏

**Pull request closed without merging**